src/essentia/essentiamath.h

```diff
@@ -882,6 +882,111 @@ TNT::Array2D<T> transpose(const TNT::Array2D<T>& m) {
   return result;
 }
 
+
+/**
+ * returns the dot product of two 1-D vectors.
+ */
+template <typename T>
+T dotProduct(const std::vector<T>& xArray, const std::vector<T>& yArray) {
+  if (xArray.empty())
+    throw EssentiaException("trying to calculate dot product of an empty array");
+  if (yArray.empty())
+    throw EssentiaException("trying to calculate dot product of an empty array");
+  return std::inner_product(xArray.begin(), xArray.end(), yArray.begin(), 0.0);
+}
+
+
+/**
+ * rotate the input 1-D vector by an index value.
+ */
+template <typename T>
+void rotateByIndex(const std::vector<T>& inputArray, const int indx) {
+  if (inputArray.empty())
+    throw EssentiaException("trying to rotate an empty array");
+  std::rotate(inputArray.begin(), inputArray.end() - indx, inputArray.end());
+}
+
+
+// returns the n-th percentile of an input array
+template <typename T> T percentile(const std::vector<T>& array, int npercentile) {
+  if (array.empty())
+    throw EssentiaException("trying to calculate percentile of empty array");
+
+  // median has sense only on sorted array
+  std::vector<T> sorted_array = array;
+  std::sort(sorted_array.begin(), sorted_array.end());
+
+  uint size = sorted_array.size()+1;
+  float nth = npercentile/100.;
+  int idx = size*nth;
```

Comment:
Is truncation intentional? We should check the specs whether to truncate or round to compute the percentiles correctly (check how it's done in scipy).